### PR TITLE
Feature/embed comments

### PR DESCRIPTION
**Pull request by nistormihai**

### CODE REVIEW

File: client/app/scripts/liveblog-edit/styles/post.less [hunk 1]
```diff
@@ -1,6 +1,11 @@
 @import 'mixins.less';
 @import 'variables.less';
 
+.lb-mainitem-comment .main-avatar, .lb-item-comment .avatar{
+    background-image: url("../../images/comments-timeline.svg") !important;
```
Comment (vied12) [inline client/app/scripts/liveblog-edit/styles/post.less]: Can't we avoid these `!important` ?
You can also play with selector deepness

File: server/liveblog/items/items.py [hunk 2]
```diff
@@ -50,6 +50,17 @@ class ItemsResource(ArchiveResource):
         'item_type': {
             'type': 'string'
         },
+        'name': {
+            'type': 'string',
+            'minlength': 1,
+            'maxlength': 300
+        },
+        'recaptcha_challenge_field': {
```
Comment (vied12) [inline server/liveblog/items/items.py]: I don't think we need these fields any more

File: server/requirements.txt [hunk 3]
```diff
@@ -5,3 +5,4 @@ git+git://github.com/joehand/flask-s3@8fe9f27ebb5251b7cf53a6684595d41dc4af2fbf#e
 # Superdesk
 -e git+git://github.com/liveblog/superdesk-core@a117d79e7fa10c92acf17aff957d05fdf2ac3f9b#egg=Superdesk-Core==0.0.1-dev
 -e git+git://github.com/liveblog/superdesk@33029d9fd55f7c7047a60abdac1f1911f044a27a#egg=Superdesk-Server
+recaptcha-client==1.0.6
```
Comment (vied12) [inline server/requirements.txt]: to be removed

File: server/settings.py [hunk 4]
```diff
@@ -185,6 +185,9 @@ def env(variable, fallback_value=None):
 S3_USE_HTTPS = False
 FLASK_ASSETS_USE_S3 = False
 USE_S3 = FLASK_ASSETS_USE_S3
+RECAPTCHA_PUBLIC_KEY = ''
```
Comment (vied12) [inline server/settings.py]: to be removed

File: server/liveblog/items/items.py (hunk 2 above)
Comment (ancafarcas) [inline server/liveblog/items/items.py]: some files are already updated regarding back-end part, in the pr #269

File: server/liveblog/themes/themes_assets/classic/main.js [hunk 5]
```diff
@@ -32,6 +33,30 @@
             loading: true,
             finished: false,
             showAuthor: SHOW_AUTHOR,
+            comment: {
```
Comment (vied12) [inline server/liveblog/themes/themes_assets/classic/main.js]: This timeline controller is already quite big, I would have created a special controller for Comments, with also a new directive to combine controller and view. What do you think?

**Comment by vied12** [comment]
LGTM, more comments in code would be awesome :relaxed:

**Comment by vied12** [comment]
And what about e2e tests ?

File: server/liveblog/themes/themes_assets/angular/liveblog-embed/resources.service.js [hunk 6]
```diff
@@ -40,20 +40,44 @@
 
     Posts.$inject = ['$resource', 'config', 'users'];
     function Posts($resource, config, users) {
+        function _completeUser(obj) {
+            if (obj.name) {
+                obj.original_creator = {display_name: obj.name};
+            } else if(obj.original_creator !== "" && obj.original_creator !== 'None'){
```
Comment (vied12) [inline server/liveblog/themes/themes_assets/angular/liveblog-embed/resources.service.js]: Why `None`? this should be `null` or `underfined`, but not `None` right ?
And no double quote in `obj.original_creator !== ""` please

File: client/app/scripts/liveblog-edit/posts.service.js [hunk 7]
```diff
@@ -93,10 +102,24 @@ define([
...
-            // complete Post With User Information
-            userList.getUser(post.original_creator).then(function(user) {
-                post.user = user;
+            post.comments = _.reduce(post.groups[1].refs, function(is, val) {
+                return is || _.isUndefined(val.item.name);
+            }, false);
+            // check if `fullDetails` flag is needed
+            // comments items set falg to true.
+            post.fullDetails = post.comments;
+            // special cases for comments.
```
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: I don't think this is the best solution. I don't see why we do need a special case for comment.
The simplest rule should be: we show the updated date when the updated date is older than creation date.
The solution would be to prevent an updated date to appear when you add a comment to a post

File: client/app/scripts/liveblog-edit/posts.service.js [hunk 8]
```diff
@@ -93,10 +102,24 @@ define([
...
             });
-            // complete Post With User Information
-            userList.getUser(post.original_creator).then(function(user) {
-                post.user = user;
+            post.comments = _.reduce(post.groups[1].refs, function(is, val) {
+                return is || _.isUndefined(val.item.name);
+            }, false);
+            // check if `fullDetails` flag is needed
+            // comments items set falg to true.
+            post.fullDetails = post.comments;
```
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: this is not clear why it is for.
This look again like a special case for comment :)
What problem does it solves ?

File: client/app/scripts/liveblog-edit/styles/post.less [hunk 9]
```diff
@@ -1,6 +1,20 @@
 @import 'mixins.less';
 @import 'variables.less';
 
+.avatar-comment {
```
Comment (vied12) [inline client/app/scripts/liveblog-edit/styles/post.less]: can it be included in a scope ? `.lb-post` maybe ?

File: server/liveblog/blogs/blogs.py [hunk 10]
```diff
@@ -147,16 +147,18 @@ def on_created(self, docs):
...
-            members = [str(m['user']) for m in doc.get('members', [])]
-            # add owner id to members
-            members.append(doc.get('original_creator'))
-            # check if current user belongs to members, and raise an exeption if not
-            if str(get_user().get('_id')) not in members:
-                roles = get_resource_service('roles').find_one(req=None, _id=get_user().get('role'))
-                if not roles:
-                    raise SuperdeskApiError.forbiddenError(message='you do not have permission to open this blog')
+        user = get_user()
+        if user:
```
Comment (vied12) [inline server/liveblog/blogs/blogs.py]: Why did you add this check?
Why a no authenticated use should access comments ?

File: server/liveblog/items/items.py [hunk 11]
```diff
@@ -50,6 +50,16 @@ class ItemsResource(ArchiveResource):
         'item_type': {
             'type': 'string'
         },
+        'name': {
+            'type': 'string',
+            'minlength': 1,
+            'maxlength': 30
+        },
+        'content': {
```
Comment (vied12) [inline server/liveblog/items/items.py]: what is `content`? what the difference with `text` ?

File: client/app/scripts/liveblog-edit/posts.service.js [hunk 12]
```diff
@@ -157,7 +180,8 @@ define([
                             blog: blog_id,
                             text: item.text,
                             meta: item.meta,
-                            item_type: item.item_type
+                            item_type: item.item_type,
+                            name: item.meta && item.meta.name
```
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: `name` can be equals to `false` here. Is it possible in the model point of view ? It seems unnatural :)

File: server/liveblog/themes/themes_assets/classic/template.html [hunk 13]
```diff
@@ -72,8 +83,42 @@ <h1 ng-if="timeline.settings.showTitle" ng-bind="timeline.blog.title"></h1>
...
+                            <div ng-if="!timeline.settings.authorNameLinksToEmail || post.mainItem.item_type === 'comment'"
+                                 class="lb-author__name"
+                                 ng-bind="post.mainItem.original_creator[timeline.settings.authorNameFormat]"></div>
+                            <div ng-if="timeline.settings.authorNameLinksToEmail && post.mainItem.item_type !== 'comment'"
+                                 class="lb-author__name">
+                                <a href="mailto:{{post.mainItem.original_creator.email}}"
+                                   ng-bind="post.mainItem.original_creator[timeline.settings.authorNameFormat]"></a>
+                            </div>
+                        </div>
+                        <div class="lb-item" ng-switch="item.item_type">
```
Comment (vied12) [inline server/liveblog/themes/themes_assets/classic/template.html]: There is too much code duplication here. I think that we can factorize it with `<script/>' templates or directives.

File: server/liveblog/themes/themes_assets/classic/template.html [hunk 14]
```diff
@@ -85,7 +130,8 @@ <h1 ng-if="timeline.settings.showTitle" ng-bind="timeline.blog.title"></h1>
                     <article ng-switch-default lb-bind-html html-content="{{ item.text }}"></article>
                 </div>
             </div>
-            <div class="lb-updated-date ng-cloak" ng-show="post._updated !== post.published_date">
+
+            <div class="lb-updated-date" ng-show="post.showUpdate">
```
Comment (vied12) [inline server/liveblog/themes/themes_assets/classic/template.html]: did you removed `ng-cloak` on purpose ?

Comment (nistormihai) [inline server/liveblog/themes/themes_assets/classic/template.html]: merge issues

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 8 above)
Comment (nistormihai) [inline client/app/scripts/liveblog-edit/posts.service.js]: `fullDetails` it is used to check if the post needs information on item.
`comments` it is used to see if the post has comments on it.

`comments` flag it is also used to compile if it is needed the *update time*

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 7 above)
Comment (nistormihai) [inline client/app/scripts/liveblog-edit/posts.service.js]: it is specified in the task.

File: server/liveblog/themes/themes_assets/angular/liveblog-embed/resources.service.js (hunk 6 above)
Comment (nistormihai) [inline server/liveblog/themes/themes_assets/angular/liveblog-embed/resources.service.js]: it comes from backend in string `None`.

File: client/app/scripts/liveblog-edit/styles/post.less (hunk 9 above)
Comment (nistormihai) [inline client/app/scripts/liveblog-edit/styles/post.less]: yes.

File: server/liveblog/blogs/blogs.py (hunk 10 above)
Comment (nistormihai) [inline server/liveblog/blogs/blogs.py]: @ancafarcas can you check this?

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 7 above)
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: Which task are you talking about ? Removing the date for post that contains a comment?
I'm talking about this one too, but I am suggesting another way to solve it, without making a special case here for comments. We agree on the problem I guess. It's about the solution here :)

File: server/liveblog/themes/themes_assets/angular/liveblog-embed/resources.service.js (hunk 6 above)
Comment (vied12) [inline server/liveblog/themes/themes_assets/angular/liveblog-embed/resources.service.js]: but this is not fine, isn't it ?

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 8 above)
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: ok then I don't think it belongs to the post service if it's related to how we want to show it in a view.
Why not putting this logic into the view and showing comments details there ? Just like we can have different template for embedded code ?

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 7 above)
Comment (nistormihai) [inline client/app/scripts/liveblog-edit/posts.service.js]: * if we publish a comment as a single-item post (comment-only) we
   * show the time of publishing (not comment creation)
   * and do not show the updated time when it is first published
* if we publish a comment in a multi-item post (comment + reaction from editors)
   * show time for each item
   * and not the updated time

Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: I agree with the finish. I'm just talking about the code. I find it strange to have a special case for comments in the posts service because of our view needs. This logic should belongs to the view from my point of view.

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 8 above)
Comment (nistormihai) [inline client/app/scripts/liveblog-edit/posts.service.js]: Business logic should belongs to the model from my point of view.

File: server/liveblog/themes/themes_assets/classic/template.html (hunk 13 above)
Comment (nistormihai) [inline server/liveblog/themes/themes_assets/classic/template.html]: I will take a look how this will be, and I think it makes sense.

File: client/app/scripts/liveblog-edit/posts.service.js [hunk 15]
```diff
@@ -93,10 +102,24 @@ define([
                 mainItem: post.groups[1].refs[0],
                 items: post.groups[1].refs
             });
-            // complete Post With User Information
-            userList.getUser(post.original_creator).then(function(user) {
-                post.user = user;
+            post.comments = _.reduce(post.groups[1].refs, function(is, val) {
+                return is || _.isUndefined(val.item.name);
```
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: post gets comments if `val.item.name` is undefined for one of the item? I don't follow the logic :)
What is "name"? Can this expression be more clear ?

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 8 above)
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: Ok if this is business logic. Then you probably need to document these fields better and the names should be more explicit (`comments` here suggest a list but this is actually a boolean and the purpose of `fullDetails` is not obvious here).
Also there is a `angular.extends` block just above. I think you should move these new field attributions into it.

File: server/liveblog/items/items.py (hunk 11 above)
Comment (ancafarcas) [inline server/liveblog/items/items.py]: the `content` has the same role as text but here we have `maxlength` and `minlength` constraints

File: server/liveblog/blogs/blogs.py (hunk 10 above)
Comment (ancafarcas) [inline server/liveblog/blogs/blogs.py]: If there is a not authenticated user, the check-ups for `admin` and `members` of the blog will not take place and they will be able to access the comments since in the comments part we used client_blogs

File: server/liveblog/items/items.py (hunk 11 above)
Comment (vied12) [inline server/liveblog/items/items.py]: So it's not good for consistency :) We can't have two different fields for the same purpose in _Items_. You should do your constraints verification at the comment level. Or to create a sub model for comments based on _Item_

File: server/liveblog/blogs/blogs.py (hunk 10 above)
Comment (vied12) [inline server/liveblog/blogs/blogs.py]: sorry, I don't understand. Does an unauthenticated user can find a comment ? Because he shouldn’t. So the remaining question is still: why did you do that ? :)

Comment (ancafarcas) [inline server/liveblog/blogs/blogs.py]: When a user is not authenticated then the `if is admin/member` check-up or if useless so I put that additional `if user` in order to give the users without credentials, possibility to post a comment

Comment (vied12) [inline server/liveblog/blogs/blogs.py]: I understand that unauthenticated users must be able to post a comment, but here we are in `find_one` method. Why a user without credentials should be allowed to find a comment ?

File: server/liveblog/items/items.py (hunk 11 above)
Comment (ancafarcas) [inline server/liveblog/items/items.py]: I moved those two fields (name and content) inside `client_modules` module so that they are comments specific now

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 8 above)
Comment (nistormihai) [inline client/app/scripts/liveblog-edit/posts.service.js]: `fullDetails` is obvious is full details on a post.
for `comments` I will change to `hasComments`.
`angular.extend` will not work here, and is a matter of preferences.

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 12 above)
Comment (nistormihai) [inline client/app/scripts/liveblog-edit/posts.service.js]: well actually is `undefined` which is how it supposed to be.

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 8 above)
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: At this logic level, `fullDetails` is not meaning full. Does it means that all the details are filled ? Are they required ? Must they be shown ? And what are details?
Regarding `angular.extend`, it's a matter of preference, but here it has been already used above, so it would be more consistent to continue like that. It also prevent to assign a value twice (jshint will ring the bell)

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 12 above)
Comment (vied12) [inline client/app/scripts/liveblog-edit/posts.service.js]: :ok_hand:

File: server/liveblog/blogs/blogs.py (hunk 10 above)
Comment (ancafarcas) [inline server/liveblog/blogs/blogs.py]: The `find_one` method is for finding a blog not a comment.

File: client/app/scripts/liveblog-edit/posts.service.js (hunk 8 above)
Comment (nistormihai) [inline client/app/scripts/liveblog-edit/posts.service.js]: let's slack this.

File: server/liveblog/blogs/blogs.py (hunk 10 above)
Comment (vied12) [inline server/liveblog/blogs/blogs.py]: right, this is for blog. But still, when I read this block, it looks like a security breach :)
I still don't understand why an unauthenticated user could find a blog when an authenticated user without permission can't do it

Comment (ancafarcas) [inline server/liveblog/blogs/blogs.py]: I guess I couldn't find another way at that moment but now...if I think better, instead of `'blog': Resource.rel('blogs', True)` I should use `'blog': Resource.rel('client_blogs', True)` and it won't be needing a security hack
